src/Storages/StorageURL.cpp

```diff
@@ -858,6 +863,54 @@ FormatSettings StorageURL::getFormatSettingsFromArgs(const StorageFactory::Argum
     return format_settings;
 }
 
+ASTs::iterator StorageURL::collectHeaders(
+    ASTs & url_function_args, URLBasedDataSourceConfiguration & configuration, ContextPtr context)
+{
+    ASTs::iterator headers_it;
+
+    for (auto arg_it = url_function_args.begin(); arg_it != url_function_args.end(); ++arg_it)
+    {
+        const auto * headers_ast_function = (*arg_it)->as<ASTFunction>();
+        if (headers_ast_function && headers_ast_function->name == "headers")
+        {
+            if (headers_it != ASTs::iterator())
+                throw Exception(
+                    ErrorCodes::BAD_ARGUMENTS,
+                    "URL table function can have only one key-value argument: headers=(). {}",
+                    bad_arguments_error_message);
+
+            const auto * headers_function_args_expr = assert_cast<const ASTExpressionList *>(headers_ast_function->arguments.get());
+            auto headers_function_args = headers_function_args_expr->children;
+
+            for (auto & header_arg : headers_function_args)
+            {
+                const auto * header_ast = header_arg->as<ASTFunction>();
+                if (!header_ast)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Headers argument is incorrect. {}", bad_arguments_error_message);
+
+                const auto * header_args_expr = assert_cast<const ASTExpressionList *>(header_ast->arguments.get());
+                auto header_args = header_args_expr->children;
+                if (header_args.size() != 2)
+                    throw Exception(ErrorCodes::BAD_ARGUMENTS, "Headers argument is incorrect. {}", bad_arguments_error_message);
+
+                auto ast_literal = evaluateConstantExpressionOrIdentifierAsLiteral(header_args[0], context);
+                auto arg_name = ast_literal->as<ASTLiteral>()->value.safeGet<String>();
```

Comment:
Let's also check that literal type is `String`, because queries like
`select * from url(..., headers('X-ClickHouse-Format'=10));`
Will produce error messages like `DB::Exception: Bad get: has UInt64, requested String. (BAD_GET)` and it's not obvious for user what is wrong